middleware/hosts/hosts.go

```diff
@@ -0,0 +1,104 @@
+package hosts
+
+import (
+	"errors"
+	"net"
+
+	"golang.org/x/net/context"
+
+	"github.com/coredns/coredns/middleware"
+	"github.com/coredns/coredns/middleware/metrics"
+	"github.com/coredns/coredns/middleware/pkg/dnsutil"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+)
+
+type (
+	Hosts struct {
+		Next middleware.Handler
+		*Hostsfile
+
+		metrics *metrics.Metrics
+	}
+)
+
+func (h Hosts) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	if state.QClass() != dns.ClassINET {
+		return dns.RcodeServerFailure, middleware.Error(h.Name(), errors.New("can only deal with ClassINET"))
+	}
+	qname := state.Name()
+
+	// Now the real zone.
+	zone := middleware.Zones(h.Names()).Matches(qname)
+
+	answers := []dns.RR{}
+
+	switch state.QType() {
+	case dns.TypePTR:
+		names := h.LookupStaticAddr(dnsutil.ExtractAddressFromReverse(zone))
+		if len(names) == 0 {
+			return middleware.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+		}
+		for _, n := range names {
+			r := new(dns.PTR)
+			r.Hdr = dns.RR_Header{Name: zone, Rrtype: dns.TypePTR,
+				Class: dns.ClassINET, Ttl: 3600}
+			r.Ptr = n
+			answers = append(answers, r)
+		}
+	case dns.TypeA:
```

Comment:
these address cases do a lot of the same work and this can probably be refactored a little.